src/content-claims/content-claims-blockstore.js

```diff
@@ -0,0 +1,379 @@
+  serviceURL
+) {
+  const claims = await read(link, { serviceURL })
+  // @todo consider checking the claims themselves
+  return Boolean(claims.length)
+}
+
+/**
+ * @param {Map<string, Promise<IndexEntry|undefined>>} map - keys are CID strings
+ * @returns {import('../api.js').IndexMap & Index}
+ */
+function createIndexEntryMap (map = new Map()) {
+  return {
+    get: async (key) => {
+      return map.get(key.toString())
+    },
+    has: async (key) => {
+      const hasKey = map.has(key.toString())
+      return hasKey
+    },
+    set: async (key, value) => {
+      map.set(key.toString(), Promise.resolve(value))
+    }
+  }
+}
+
+/**
+ * get index for a link from a content claims client
+ * @param {AsyncIterable<import('../api.js').LocationClaim>|Iterable<import('../api.js').LocationClaim>} claims
+ * @param {UnknownLink} link - link to answer whether the content-claims at `url` has blocks for `link`
+ * @returns {AsyncGenerator<Uint8Array>}
+ */
+async function * fetchBlocksForLocationClaims (claims, link) {
+  /**
+   * @type {ReadableStream<{
+   *   claim: import('@web3-storage/content-claims/client/api').LocationClaim
+   *   location: string
+   *   response: Response
+   *   block: import('carstream/api').Block
+   * }>}
+   */
+  const locatedBlocks = createReadableStream(claims)
+    // claim -> [claim, location]
+    .pipeThrough(new TransformStream({
+    /**
+      * @param {import('@web3-storage/content-claims/client/api').LocationClaim} claim
+      */
+      async transform (claim, controller) {
+        for (const location of claim.location) { controller.enqueue([claim, location]) }
+      }
+    }))
+    // [claim, location] -> [claim, location, ok response]
+    .pipeThrough(new TransformStream({
+    /** @param {[claim: import('@web3-storage/content-claims/client/api').Claim, location: string]} chunk */
+      async transform ([claim, location], controller) {
+        const response = await fetch(location, { headers: { accept: 'application/vnd.ipld.car' } })
+        if (response.ok) {
+          return controller.enqueue([claim, location, response])
+        }
+        throw new Error(`unexpected not-ok response ${response}`)
+      }
+    }))
+    // [claim, location, ok response] -> { claim, location, ok response, block }
+    .pipeThrough(new TransformStream({
+    /** @param {[claim: import('@web3-storage/content-claims/client/api').LocationClaim, location: string, response: Response]} chunk */
+      async transform ([claim, location, response], controller) {
+        await response.body?.pipeThrough(new CARReaderStream()).pipeTo(new WritableStream({
+          write: (block) => {
+          // @todo validate that block.bytes hashes to `link`
+            controller.enqueue({ claim, location, response, block })
+          }
+        }))
+      }
+    }))
+  for await (const locatedBlock of locatedBlocks) {
+    yield locatedBlock.block.bytes
+  }
+}
+
+/**
+ * @param {import('../api.js').RelationClaim[]} claims
```

Comment:
would love a good comment here summarizing what's going on - there's some complicated logic below and I'd love an abstract